src/lower/mod.rs

```diff
@@ -791,14 +787,19 @@ impl ir::ImplDatum {
     /// Given `impl<T: Clone> Clone for Vec<T>`, generate:
     ///
     /// ```notrust
-    /// forall<T> { (Vec<T>: Clone) :- (T: Clone) }
+    /// forall<T> { (Vec<T>: Clone) :- (T: Clone), WF(T: Clone) }
     /// ```
-    fn to_program_clause(&self) -> ir::ProgramClause {
+    fn to_program_clause(&self, program: &ir::Program) -> ir::ProgramClause {
         ir::ProgramClause {
             implication: self.binders.map_ref(|bound| {
                 ir::ProgramClauseImplication {
                     consequence: bound.trait_ref.clone().cast(),
-                    conditions: bound.where_clauses.clone().cast(),
+                    conditions: bound.where_clauses
+                                     .iter()
+                                     .cloned()
+                                     .flat_map(|wc| wc.expanded(program))
+                                     .map(|wc| wc.cast())
```

Comment:
apropos of nothing, we should make a `.casted()` adapter just to make this cuter :)